File: BCycle/BCycleStationsViewController.m
//
//  BCycleStationsViewController.m
//  BCycle
//
//  Created by Thomas Traylor on 9/22/14.
//  Copyright (c) 2014 Thomas Traylor. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import <MapKit/MapKit.h>
#import "BCycleStationsViewController.h"
#import "BCycleServices.h"
#import "StationInformationViewController.h"

#if !defined(METERS_PER_MILE)
#define METERS_PER_MILE 1609.344
#endif

@interface BCycleStationsViewController ()


@property (nonatomic, weak) IBOutlet MKMapView *mapView;
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, strong) CLLocation *currentLocation;
@property (nonatomic, strong) CLPlacemark *currentPlacemark;
@property (nonatomic, strong) BCycleServices *bCycleServices;
@property (nonatomic) BOOL userInteractionCausedRegionChange;
@property (nonatomic) MKMapRect currentSearchRect;
@property (nonatomic) MKCoordinateSpan currentSearchSpan;

- (void)addAnnotationsForRegion:(MKCoordinateRegion)region;
- (MKMapRect)createRectForRegion:(MKCoordinateRegion)region;
- (MKCoordinateRegion)createSearchRegionForLocation:(CLLocationCoordinate2D)coordinate andDistance:(CLLocationDistance)distance;
- (MKCoordinateRegion)createViewableRegionForLocation:(CLLocationCoordinate2D)coordinate andDistance:(CLLocationDistance)distance;
- (MKCoordinateRegion)createNewSearchRegionForRegion:(MKCoordinateRegion)region;

@end

@implementation BCycleStationsViewController

#pragma mark - Setters

- (CLLocationManager*)locationManager {
    if (_locationManager == nil) {
        _locationManager = [[CLLocationManager alloc] init];
        _locationManager.delegate = self;
        _locationManager.distanceFilter = kCLDistanceFilterNone;
        _locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters;
        _locationManager.activityType = CLActivityTypeOtherNavigation;
    }
    
    return _locationManager;
}

- (BCycleServices*)bCycleServices {
    if (_bCycleServices == nil) {
        _bCycleServices = [[BCycleServices alloc] init];
    }
    
    return _bCycleServices;
}

#pragma  mark - ViewController Life Cycle

- (void)viewDidLoad {
    
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.title = @"BCycle";
    
    self.userInteractionCausedRegionChange = NO;
    self.mapView.delegate = self;
    self.mapView.showsUserLocation = YES;
    self.mapView.pitchEnabled = YES;
    self.mapView.zoomEnabled = YES;
    self.mapView.rotateEnabled = YES;
    self.navigationController.navigationBar.tintColor = [UIColor redColor];
    
    NSDictionary *textAttributes = @{NSForegroundColorAttributeName : [UIColor redColor]};
    self.navigationController.navigationBar.titleTextAttributes = textAttributes;

    [self.locationManager requestWhenInUseAuthorization];
    [self.locationManager startUpdatingLocation];
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations {
    self.currentLocation = [locations lastObject];
    
    // once we know where we are, we don't need to keep the location services running,
    // so stop it
    [self.locationManager stopUpdatingLocation];
    
    // save our current Placemark.
    CLGeocoder *geoCoder = [[CLGeocoder alloc] init];
    [geoCoder reverseGeocodeLocation:self.currentLocation completionHandler:^(NSArray *placemarks, NSError *error) {
        if(error == nil) {
            self.currentPlacemark = placemarks[0];
        }
    }];
    
    // remove the annotations if there are any
    [self.mapView removeAnnotations:self.mapView.annotations];
    // center the map
    [self.mapView setCenterCoordinate:self.currentLocation.coordinate animated:YES];
    
    // create the initial search area
    MKCoordinateRegion searchRegion = [self createSearchRegionForLocation:self.currentLocation.coordinate andDistance:1.0f];
    
    // fetch the data for the annotations
    [self addAnnotationsForRegion:searchRegion];
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    
    if(error)
        NSLog(@"[%@ %@] error: %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd), [error localizedDescription]);
    
    if ([error code] != kCLErrorLocationUnknown) {
        [self.locationManager stopUpdatingLocation];
        // since we only have data for Denver, we'll center the map there
        self.mapView.centerCoordinate = CLLocationCoordinate2DMake(39.758202, -105.001359);
    }
    
}

-(void)locationManagerDidChangeAuthorization:(CLLocationManager *)manager {
    
    switch (self.locationManager.authorizationStatus) {
            
        case kCLAuthorizationStatusNotDetermined:
        case kCLAuthorizationStatusAuthorizedAlways:
        case kCLAuthorizationStatusAuthorizedWhenInUse:
            [self.locationManager startUpdatingLocation];
            break;
        case kCLAuthorizationStatusRestricted:
        case kCLAuthorizationStatusDenied:
            [self.locationManager stopUpdatingLocation];
            break;
            
    }
}

#pragma mark - MKMapViewDelegate

- (void)mapView:(MKMapView *)mapView didUpdateUserLocation:(MKUserLocation *)userLocation {
    MKCoordinateRegion region = [self createViewableRegionForLocation:userLocation.coordinate andDistance:0.5f];
    if (self.userInteractionCausedRegionChange == NO) {
        [self addAnnotationsForRegion:region];
    }
    [self.mapView setRegion:region animated:YES];
}

- (void)mapView:(MKMapView *)mapView regionWillChangeAnimated:(BOOL)animated {
    UIView* view = mapView.subviews.firstObject;
    // check to see if the user is interacting with the map
    for(UIGestureRecognizer* recognizer in view.gestureRecognizers) {
        if(recognizer.state == UIGestureRecognizerStateBegan || recognizer.state == UIGestureRecognizerStateEnded) {
            self.userInteractionCausedRegionChange = YES;
            break;
        }
    }
}

-(void)mapView:(MKMapView *)mapView regionDidChangeAnimated:(BOOL)animated {
    if(self.userInteractionCausedRegionChange) {
        self.userInteractionCausedRegionChange = NO;
        MKMapPoint mapPoint = MKMapPointForCoordinate(mapView.region.center);
        if (MKMapRectContainsPoint(self.currentSearchRect, mapPoint)) {
            MKMapRect viewableRect = [self createRectForRegion:mapView.region];
            if (!MKMapRectContainsRect(self.currentSearchRect, viewableRect)) {
                // the user has zoomed out but the viewable region's center point is still
                // within the current search map rect. We will use the current region center
                // and increase the span by some percentage to create a new search region (map rect)
                MKCoordinateRegion region = [self createNewSearchRegionForRegion:mapView.region];
                [self addAnnotationsForRegion:region];
            }
        }
        else {
            // the user may have zoomed out, but nonetheless the region center is outside the
            // current search map rect. We will use the current region center and increase
            // the span by some percentage to create a new search region (map rect)
            MKCoordinateRegion region = [self createNewSearchRegionForRegion:mapView.region];
            [self addAnnotationsForRegion:region];
        }
    }
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
    // the mkannotationview we'll return
    MKAnnotationView *view = nil;
    
    //if the annotation is a user location
    if ([annotation isKindOfClass:[MKUserLocation class]]) {
        view = nil;
    }
    else {
        view = [mapView dequeueReusableAnnotationViewWithIdentifier:NSStringFromClass([annotation class])];
        
        if(!view) {
            MKMarkerAnnotationView *marker = [[MKMarkerAnnotationView alloc] initWithAnnotation:annotation
                                                                             reuseIdentifier:NSStringFromClass([annotation class])];
            marker.canShowCallout = YES;
            marker.enabled = YES;
            marker.draggable = NO;
            marker.markerTintColor = [UIColor redColor];
            
            if([mapView.delegate respondsToSelector:@selector(mapView:annotationView:calloutAccessoryControlTapped:)]) {
                marker.rightCalloutAccessoryView = [UIButton buttonWithType:UIButtonTypeDetailDisclosure];
            }
            
            view = marker;
        }
        else {
            // set the annotation for the view
            view.annotation = annotation;
        }
    }
    
    return view;
}

- (void)mapView:(MKMapView *)mapView annotationView:(MKAnnotationView *)view calloutAccessoryControlTapped:(UIControl *)control {
    [self performSegueWithIdentifier:@"StationInformation" sender:view];
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"StationInformation"]) {
        if ([sender isKindOfClass:[MKAnnotationView class]]) {
            MKAnnotationView *aView = sender;
            if ([aView.annotation isKindOfClass:[BCycleStation class]]) {
                BCycleStation *station = (id) aView.annotation;
                if ([segue.destinationViewController respondsToSelector:@selector(setStation:)]) {
                    [segue.destinationViewController performSelector:@selector(setStation:)
                                                          withObject:station];
                }
            }
        }
    }
    else if ([segue.identifier isEqualToString:@"AddStation"]) {
        if ([segue.destinationViewController respondsToSelector:@selector(setCurrentPlacemark:)]) {
            [segue.destinationViewController performSelector:@selector(setCurrentPlacemark:)
                                                  withObject:self.currentPlacemark];
        }
    }
}

#pragma mark - Private Methods

- (void)addAnnotationsForRegion:(MKCoordinateRegion)region
{
    [self.bCycleServices getStationsInRegion:region withCompletion:^(NSArray *result, NSError *error) {
        
        if(result.count > 0) {
            dispatch_queue_t geoQue = dispatch_queue_create("geoQue", (dispatch_queue_attr_t) 0);
            dispatch_async(geoQue, ^{
                
                NSMutableArray *bikes = [[NSMutableArray alloc] init];
                for (NSDictionary *station in result) {
                    CLLocationDegrees latitude = [[station valueForKey:@"latitude"] floatValue];
                    CLLocationDegrees longitude = [[station valueForKey:@"longitude"] floatValue];
                    CLLocationCoordinate2D location = CLLocationCoordinate2DMake(latitude, longitude);
                    
                    BCycleStation *bc = [[BCycleStation alloc] initWithCoordinate:location];
                    [bc setName:[station valueForKey:@"name"]];
                    [bc setStreet:[station valueForKey:@"street"]];
                    [bc setCity:[station valueForKey:@"city"]];
                    [bc setState:[station valueForKey:@"state"]];
                    [bc setZip:[station valueForKey:@"zip"]];
                    
                    [bikes addObject:bc];
                }
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    
                    [self.mapView addAnnotations:bikes];
                    
                });
            });
        }
    }];
}

#pragma mark - Map Regions

- (MKMapRect)createRectForRegion:(MKCoordinateRegion)region {
    
    MKMapPoint a = MKMapPointForCoordinate(CLLocationCoordinate2DMake(region.center.latitude + region.span.latitudeDelta / 2.0,
                                                                      region.center.longitude - region.span.longitudeDelta / 2.0));
    MKMapPoint b = MKMapPointForCoordinate(CLLocationCoordinate2DMake(region.center.latitude - region.span.latitudeDelta / 2.0,
                                                                      region.center.longitude + region.span.longitudeDelta / 2.0));
    
    return MKMapRectMake(MIN(a.x,b.x), MIN(a.y,b.y), ABS(a.x-b.x), ABS(a.y-b.y));
}

- (MKCoordinateRegion)createSearchRegionForLocation:(CLLocationCoordinate2D)coordinate andDistance:(CLLocationDistance)distance {
    CLLocationDirection latInMeters = distance*METERS_PER_MILE;
    CLLocationDirection longInMeters = distance*METERS_PER_MILE;
    MKCoordinateRegion region = MKCoordinateRegionMakeWithDistance(coordinate, latInMeters, longInMeters);
    
    self.currentSearchSpan = region.span;
    self.currentSearchRect = [self createRectForRegion:region];
    
    return region;
}

- (MKCoordinateRegion)createViewableRegionForLocation:(CLLocationCoordinate2D)coordinate andDistance:(CLLocationDistance)distance {
    CLLocationDirection latInMeters = distance*METERS_PER_MILE;
    CLLocationDirection longInMeters = distance*METERS_PER_MILE;
    MKCoordinateRegion region = MKCoordinateRegionMakeWithDistance(coordinate, latInMeters, longInMeters);
    
    return region;
}

- (MKCoordinateRegion)createNewSearchRegionForRegion:(MKCoordinateRegion)region {
    MKCoordinateSpan span = region.span;
    if(span.latitudeDelta < self.currentSearchSpan.latitudeDelta) {
        // since the viewable span's latitude delta is less than
        // the search span latitude delta we'll add the diff of the two
        // to the current viewable span
        span.latitudeDelta = self.currentSearchSpan.latitudeDelta + (self.currentSearchSpan.latitudeDelta - span.latitudeDelta);
    }
    else {
        // since the viewable span's latitude delta is larger than the search span
        // latitude delta we'll add 50% to the viewable region
        span.latitudeDelta += (span.latitudeDelta * 0.5f);
    }
    
    MKCoordinateRegion newRegion = MKCoordinateRegionMake(region.center, span);
    self.currentSearchSpan = newRegion.span;
    self.currentSearchRect = [self createRectForRegion:newRegion];
    return newRegion;
}

@end

File: BCycle/BCycleStationsViewController.swift
//
//  BCycleStationsViewController.swift
//  BCycle
//
//  Created by Thomas Traylor on 1/1/24.
//

import UIKit
import MapKit
import CoreLocation

@MainActor
class BCycleStationsViewController: UIViewController {
    
    private let MetersPerMile: Double = 1609.344
    private var currentPlacemark: CLPlacemark?
    private var currentLocation: CLLocation?
    private var locationManager: CLLocationManager = CLLocationManager()
    private var userInteractionCausedRegionChange: Bool = false
    
    private var mapView: MKMapView = {
        return MKMapView(frame: .zero)
    }()
    
    override func loadView() {
        view = UIView(frame: .zero)
        view.backgroundColor = .systemBackground
        
        view.addSubview(mapView)
        mapView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            mapView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            mapView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            view.bottomAnchor.constraint(equalTo: mapView.bottomAnchor),
            view.safeAreaLayoutGuide.trailingAnchor.constraint(equalTo: mapView.trailingAnchor)
        ])
        
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        let navBarAppearance = UINavigationBarAppearance()
        navBarAppearance.configureWithOpaqueBackground()
        navBarAppearance.titleTextAttributes = [.foregroundColor: UIColor.systemRed]
        navBarAppearance.largeTitleTextAttributes = [.foregroundColor: UIColor.systemRed]
        navigationController?.navigationBar.standardAppearance = navBarAppearance
        navigationController?.navigationBar.scrollEdgeAppearance = navBarAppearance
        navigationController?.navigationBar.standardAppearance = navBarAppearance
        navigationController?.navigationBar.compactAppearance = navBarAppearance
        navigationController?.navigationBar.scrollEdgeAppearance = navBarAppearance
        navigationController?.navigationBar.compactScrollEdgeAppearance = navBarAppearance
        navigationController?.navigationBar.prefersLargeTitles = false
        navigationController?.navigationBar.isTranslucent = false
        navigationController?.navigationBar.tintColor = .white
        
        setNeedsStatusBarAppearanceUpdate()
        title = "BCycle"
        
        let addButton = UIBarButtonItem(image: UIImage(systemName: "plus"), style: .plain, target: self, action: #selector(addStation(_:)))
        addButton.tintColor = .systemRed
        navigationItem.setRightBarButton(addButton, animated: true)
        
        mapView.delegate = self
        mapView.showsUserLocation = true
        mapView.isPitchEnabled = true
        mapView.isZoomEnabled = true
        mapView.isRotateEnabled = true
        mapView.register(
            StationAnnotationView.self,
            forAnnotationViewWithReuseIdentifier: StationAnnotationView.reuseId) // MKMapViewDefaultAnnotationViewReuseIdentifier)
        mapView.register(
            StationClusterAnnotationView.self,
            forAnnotationViewWithReuseIdentifier: MKMapViewDefaultClusterAnnotationViewReuseIdentifier)
        
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        
        let longPressGesture = UILongPressGestureRecognizer(target: self, action: #selector(dropAPinForNewStation(_:)))
        mapView.addGestureRecognizer(longPressGesture)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            let results = CLLocationManager.locationServicesEnabled()
            DispatchQueue.main.async {
                if results {
                    switch self.locationManager.authorizationStatus {
                    case .notDetermined:
                        self.locationManager.requestWhenInUseAuthorization()
                    case .restricted, .denied:
                        // show message
                        self.locationManager.stopUpdatingLocation()
                        self.locationManagerError()
                    case .authorizedWhenInUse, .authorizedAlways:
                        /// app is authorized
                        self.locationManager.requestLocation()
                    default:
                        self.locationManager.requestWhenInUseAuthorization()
                        break
                    }
                }
                else {
                    self.locationManagerError()
                }
            }
        }
    }
}

private extension BCycleStationsViewController {
    
    func locationManagerError() {
        DispatchQueue.main.async {
            let msg = "Unable to get the current location.  Using Denver, CO."
            let alert = UIAlertController(title: "Location Error",
                                          message: msg,
                                          preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in
                self.currentLocation = CLLocation(latitude: 39.758202, longitude: -105.001359)
                self.mapView.setCenter(self.currentLocation!.coordinate, animated: true)
                // create the initial search area
                let searchRegion: MKCoordinateRegion = self.createSearchRegionForLocation(self.currentLocation!.coordinate,
                                                                                          distance: 0.5)
                self.mapView.setRegion(searchRegion, animated: true)
                self.addAnnotationsForRegion(searchRegion)
                if let currentLocation = self.currentLocation {
                    CLGeocoder().reverseGeocodeLocation(currentLocation) { clPlacmarks, _ in
                        if let placemarks = clPlacmarks, let placemark = placemarks.first {
                            self.currentPlacemark = placemark
                        }
                    }
                }
            }))
            self.present(alert, animated: true)
        }
    }
    
    func addAnnotationsForRegion(_ region: MKCoordinateRegion) {
        Task {
            do {
                let bcycleApiService: BCycleApiService = BCycleApiService()
                let stations = try await bcycleApiService.getStationsInRegion(region)
                DispatchQueue.main.async {
                    for annotation in self.mapView.annotations {
                        if let station = annotation as? StationAnnotation {
                            self.mapView.removeAnnotation(station)
                        }
                        if let cluster = annotation as? MKClusterAnnotation {
                            self.mapView.removeAnnotation(cluster)
                        }
                    }
                    
                    let stationAnnotations: [StationAnnotation] = stations.map({ StationAnnotation(station: $0) })
                    self.mapView.addAnnotations(stationAnnotations)
                    self.mapView.reloadInputViews()
                }
            }
            catch {
                DispatchQueue.main.async {
                    let alertController = UIAlertController(
                        title: NSLocalizedString("No BCycle stations found.", comment: "No BCycle stations found."),
                        message: "We were unable to locate BCycle stations in your area.",
                        preferredStyle: .alert
                    )
                    
                    let continueAction = UIAlertAction(title: "OK", style: .default, handler: nil)
                    
                    alertController.addAction(continueAction)
                    self.present(alertController, animated: true, completion: nil)
                }
            }
        }
    }
    
    func createRectForRegion(_ region: MKCoordinateRegion) -> MKMapRect {
        
        let a = MKMapPoint(CLLocationCoordinate2DMake(region.center.latitude + region.span.latitudeDelta / 2.0,
                                                      region.center.longitude - region.span.longitudeDelta / 2.0))
        let b = MKMapPoint(CLLocationCoordinate2DMake(region.center.latitude - region.span.latitudeDelta / 2.0,
                                                      region.center.longitude + region.span.longitudeDelta / 2.0))
        
        return MKMapRect(x: min(a.x,b.x), y: min(a.y,b.y), width: abs(a.x-b.x), height: abs(a.y-b.y))
    }
    
    func createSearchRegionForLocation(_ coordinate: CLLocationCoordinate2D, distance: CLLocationDistance) -> MKCoordinateRegion {
        
        let latInMeters: CLLocationDirection = distance * MetersPerMile
        let longInMeters: CLLocationDirection = distance * MetersPerMile
        let region: MKCoordinateRegion = MKCoordinateRegion(center: coordinate,
                                                            latitudinalMeters: latInMeters,
                                                            longitudinalMeters: longInMeters)
        
        return region
    }
    
    func createViewableRegionForLocation(_ coordinate: CLLocationCoordinate2D, distance: CLLocationDistance) -> MKCoordinateRegion {
        let latInMeters: CLLocationDirection = distance * MetersPerMile
        let longInMeters: CLLocationDirection = distance * MetersPerMile
        return MKCoordinateRegion(center: coordinate, latitudinalMeters: latInMeters, longitudinalMeters: longInMeters)
    }
    
    
    // MARK: - Actions
    @objc func addStation(_ barButtonItem: UIBarButtonItem) {
        let viewController = AddStationViewController()
        viewController.placemark = currentPlacemark
        viewController.delegate = self
        let nav = UINavigationController(rootViewController: viewController)
        nav.modalPresentationStyle = .pageSheet
        if let sheet = nav.sheetPresentationController {
            sheet.detents = [.medium(), .large()]
            sheet.prefersGrabberVisible = true
            sheet.selectedDetentIdentifier = .large
            sheet.prefersScrollingExpandsWhenScrolledToEdge = false
        }
        present(nav, animated: true)
    }
    
    @objc func dropAPinForNewStation(_ sender: UILongPressGestureRecognizer) {
        if sender.state != UIGestureRecognizer.State.ended {
            let touchLocation = sender.location(in: self.mapView)
            let locationCoordinate = self.mapView.convert(touchLocation, toCoordinateFrom: self.mapView)
            print("Latitude: \(locationCoordinate.latitude), Longitude: \(locationCoordinate.longitude)")
            let newBcycleLocation = MKPointAnnotation()
            newBcycleLocation.coordinate = locationCoordinate
            newBcycleLocation.title = "Dropped Pin"
            newBcycleLocation.subtitle = "Latitude: \(locationCoordinate.latitude), Longitude: \(locationCoordinate.longitude)"
            self.mapView.addAnnotation(newBcycleLocation)
        }
    }
}

// MARK: - CLLocationManagerDelegate
extension BCycleStationsViewController : @preconcurrency CLLocationManagerDelegate {
    
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        switch manager.authorizationStatus {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        case .restricted, .denied:
            // show message
            manager.stopUpdatingLocation()
            locationManagerError()
        case .authorizedWhenInUse, .authorizedAlways:
            /// app is authorized
            manager.requestLocation()
        default:
            locationManagerError()
            break
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("error: \(error.localizedDescription)")
        if let clerror = error as? CLError, clerror.code != .locationUnknown {
            locationManager.stopUpdatingLocation()
            mapView.centerCoordinate = CLLocationCoordinate2DMake(39.758202, -105.001359)
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        currentLocation = locations.last
        
        if let currentLocation = currentLocation {
            locationManager.stopUpdatingLocation()
            CLGeocoder().reverseGeocodeLocation(currentLocation) { clPlacmarks, _ in
                if let placemarks = clPlacmarks, let placemark = placemarks.first {
                    self.currentPlacemark = placemark
                }
            }
        }
        
        if let coordinate = self.currentLocation?.coordinate {
            DispatchQueue.main.async {
                self.mapView.setCenter(coordinate, animated: true)
            }
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        print("location manager did change authorization status")
    }
}

// MARK: - MKMapViewDelegate
extension BCycleStationsViewController : MKMapViewDelegate {
    
    func mapView(_ mapView: MKMapView, didUpdate userLocation: MKUserLocation) {
        currentLocation = userLocation.location
        if let coordiante = userLocation.location?.coordinate {
            mapView.setCenter(coordiante, animated: true)
        }
        
        let region: MKCoordinateRegion = createViewableRegionForLocation(userLocation.coordinate, distance: 0.5)
        mapView.setRegion(region, animated: true)
        if !userInteractionCausedRegionChange {
            addAnnotationsForRegion(region)
        }
    }
    
    func mapView(_ mapView: MKMapView, regionWillChangeAnimated animated: Bool) {
        if let view = mapView.subviews.first, let gestureRecognizers = view.gestureRecognizers {
            for recognizer in gestureRecognizers {
                if recognizer.state == .began || recognizer.state == .ended || recognizer.state == .changed {
                    userInteractionCausedRegionChange = true
                    break
                }
            }
        }
    }
    
    func mapView(_ mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
        if userInteractionCausedRegionChange {
            userInteractionCausedRegionChange = false
            addAnnotationsForRegion(mapView.region)
        }
    }
    
    func mapView(_ mapView: MKMapView, didSelect view: MKAnnotationView) {
        
        guard !(view.annotation is MKUserLocation) else { return }
        
        var newViewController: UIViewController?
        
        if view.annotation is StationAnnotation {
            let stationAnnotation = view.annotation as! StationAnnotation
            let viewController = StationInformationViewController()
            viewController.title = stationAnnotation.station.name ?? "BCycle Location"
            viewController.stationAnnotation = stationAnnotation
            viewController.userLocation = currentLocation
            viewController.delegate = self
            newViewController = viewController
        }
        else if view.annotation is MKPointAnnotation {
            let viewController = DroppedPinViewController()
            viewController.title = "Dropped Pin"
            viewController.pinLocation = view.annotation?.coordinate
            viewController.annotation = view.annotation
            viewController.delegate = self
            newViewController = viewController
        }
        
        guard let newViewController = newViewController else { return }
        let nav = UINavigationController(rootViewController: newViewController)
        nav.modalPresentationStyle = .pageSheet
        if let sheet = nav.sheetPresentationController {
            sheet.detents = [.medium(), .large()]
            sheet.prefersGrabberVisible = true
            sheet.selectedDetentIdentifier = .medium
            sheet.prefersScrollingExpandsWhenScrolledToEdge = false
        }
        
        // Display the annotation information
        present(nav, animated: true)
    }
    
    func mapView(_ mapView: MKMapView, viewFor annotation: any MKAnnotation) -> MKAnnotationView? {
        
        guard let annotation = annotation as? StationAnnotation else { return nil }
        
        guard let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: StationAnnotationView.reuseId) as? StationAnnotationView else {
            return StationAnnotationView(annotation: annotation, reuseIdentifier: StationAnnotationView.reuseId)
        }
        annotationView.annotation = annotation
        annotationView.clusteringIdentifier = StationAnnotationView.clusteringId
        return annotationView
    }
    
}

// MARK: - AddStationViewControllerDelegate
extension BCycleStationsViewController: @preconcurrency AddStationViewControllerDelegate {
    
    func addStationViewController(_ controller: AddStationViewController, didAddStationAt placemark: CLPlacemark) {
        addAnnotationsForRegion(mapView.region)
    }
}

// MARK: - DropPinViewControllerDelegate
extension BCycleStationsViewController: @preconcurrency DropPinViewControllerDelegate {
    
    func droppedPinViewController(_ controller: DroppedPinViewController, remove annotation: any MKAnnotation) {
        self.mapView.removeAnnotation(annotation)
        addAnnotationsForRegion(mapView.region)
    }
    
    func droppedPinViewController(_ controller: DroppedPinViewController, didDeselect annotation: any MKAnnotation) {
        self.mapView.deselectAnnotation(annotation, animated: true)
    }
}

// MARK: - StationInformationViewControllerDelegate
extension BCycleStationsViewController: @preconcurrency StationInformationViewControllerDelegate {
    
    func stationInformationViewController(_ viewController: StationInformationViewController, didDeselect annotation: StationAnnotation) {
        self.mapView.deselectAnnotation(annotation, animated: true)
    }
    
}
